std/experimental/allocator/building_blocks/aligned_block_list.d

```diff
@@ -0,0 +1,690 @@
+// Written in the D programming language.
+/**
+Source: $(PHOBOSSRC std/experimental/allocator/building_blocks/_aligned_block_list.d)
+*/
+module std.experimental.allocator.building_blocks.aligned_block_list;
+
+import std.experimental.allocator.common;
+import std.experimental.allocator.building_blocks.null_allocator;
+
+// Common function implementation for thread local and shared AlignedBlockList
+private mixin template AlignedBlockListImpl(bool isShared)
+{
+    import std.traits : hasMember;
+    import std.typecons : Ternary;
+
+    static if (isShared)
+    import core.internal.spinlock : SpinLock;
+
+private:
+    // Doubly linked list of 'AlignedBlockNode'
+    // Each node contains an `Allocator` followed by its payload
+    static struct AlignedBlockNode
+    {
+        AlignedBlockNode* next, prev;
+        Allocator bAlloc;
+
+        static if (isShared)
+        {
+            shared(size_t) bytesUsed;
+            // Since the lock is not taken when allocating, this acts like a refcount
+            // keeping the node alive
+            uint keepAlive;
+        }
+        else
+        {
+            size_t bytesUsed;
+        }
+    }
+
+    // Root of the internal doubly linked list
+    AlignedBlockNode* root;
+
+    // Number of active nodes
+    int numNodes;
+
+    // If the numNodes exceeds this limit, we will start deallocating nodes
+    enum maxNodes = 50;
+
+    // This lock is always taken when changing the list
+    // To improve performance, the lock is not taken when the allocation logic is called
+    static if (isShared)
+    SpinLock lock = SpinLock(SpinLock.Contention.brief);
+
+    // Moves a node to the front of the list, allowing for quick allocations
+    void moveToFront(AlignedBlockNode* tmp)
+    {
+        auto localRoot = cast(AlignedBlockNode*) root;
+        if (tmp == localRoot)
+            return;
+
+        if (tmp.prev) tmp.prev.next = tmp.next;
+        if (tmp.next) tmp.next.prev = tmp.prev;
+        if (localRoot) localRoot.prev = tmp;
+        tmp.next = localRoot;
+        tmp.prev = null;
+
+        root = cast(typeof(root)) tmp;
+    }
+
+    // Removes a node from the list, including its payload
+    // The payload is deallocated by calling 'parent.deallocate'
+    void removeNode(AlignedBlockNode* tmp)
+    {
+        auto next = tmp.next;
+        if (tmp.prev) tmp.prev.next = tmp.next;
+        if (tmp.next) tmp.next.prev = tmp.prev;
+        assert(parent.deallocate((cast(void*) tmp)[0 .. alignment]));
```

Comment:
In release mode assertions are omitted and this code won't be run.

Example: https://run.dlang.io/is/m8gzG3
```d
import std.stdio;
auto foo(ref int x)
{
    return x &= 1;
}
void main(string[] args)
{
    int x = 255;
    assert(foo(x));
    writeln(x);
}
```

This prints 1 when run with no compiler arguments but prints 255 with `-release`.

Reply:
Thanks, good catch!